exla/lib/exla/defn.ex

```diff
@@ -541,15 +541,39 @@ defmodule EXLA.Defn do
     {fun_computation(name, args, expr, type, state), cache}
   end
 
-  defp cached_recur_operator(:optional, %T{data: %Expr{args: [expr, default]}}, state, cache) do
-    {args, cache} =
+  defp cached_recur_operator(:optional, %T{data: %Expr{args: args}}, state, cache) do
+    [expr, default_body] = args
+
+    computation_args =
       expr.data.args
       |> Enum.take_while(&(not is_list(&1)))
-      |> Enum.map_reduce(cache, &recur_operator(&1, state, &2))
 
-    params = Enum.with_index(args, fn arg, index -> {index, arg} end)
-    state = %{state | params: Map.new(params), scope_ids: Tree.scope_ids(default)}
-    recur_operator(default, state, cache)
+    {call_args, cache} =
+      computation_args |> Enum.map_reduce(cache, &recur_operator(&1, state, &2))
+
+    token = get_token(cache)
+
+    {call_body, cache} =
+      optional_computation(
+        :optional_body,
+        List.to_tuple(computation_args),
+        default_body,
+        state,
+        cache
+      )
+
+    builder =
+      case call_args do
+        [%{builder: builder} | _] ->
+          builder
+
+        _ ->
+          state.builder
+      end
+
+    call = EXLA.Op.call(builder, token, call_args, call_body)
```

Comment:
We should pass the token as a regular argument (iirc the first argument) in `call_args`, no?

Reply:
I.e. EXLA.Op.call(builder, computation, args)